[DPE-5965] - Set private key with juju secrets

skourta opened the pull request:
This pull request introduces the ability to set TLS private keys through a config option to pass a juju secret id.
The TLS v4 lib used is forked [here](https://github.com/skourta/tls-certificates-interface/tree/private-key-secret-label).
### Config and literals:
* [`config.yaml`](diffhunk://#diff-d8d0422389f03d783e32e627250fe29834bd09c6361640d1ff00661dd6820034R12-R17): Added a new configuration option `tls-private-key` for specifying the private key for TLS certificates.
* [`src/literals.py`](diffhunk://#diff-959ba637043ab4ac88c6f9f9109022f78a7df1b9a2ab3a0407bff4f124c2db64R36): Introduced `TLS_PRIVATE_KEY_CONFIG` constant to reference the new configuration option.

### Event Handling Updates:
* [`src/events/etcd.py`](diffhunk://#diff-4d4b45129d72bf79a08e70009a59f4ebe0a016b4727117155249c864ae4a1b49R125-R132): Updated event handlers to manage the TLS private key, including fetching and updating the private key from secrets. [[1]](diffhunk://#diff-4d4b45129d72bf79a08e70009a59f4ebe0a016b4727117155249c864ae4a1b49R125-R132) [[2]](diffhunk://#diff-4d4b45129d72bf79a08e70009a59f4ebe0a016b4727117155249c864ae4a1b49R181-R190) [[3]](diffhunk://#diff-4d4b45129d72bf79a08e70009a59f4ebe0a016b4727117155249c864ae4a1b49R214-R245)

### Testing Improvements:
* [`tests/integration/helpers.py`](diffhunk://#diff-e9349e06331f9876f984bcf05671235cfbe546046bccab6ad1abb0b183982b30R129-R149): Added a helper function `add_secret` to facilitate adding secrets during integration tests.
* [`tests/integration/tls/test_private_key_rotation.py`](diffhunk://#diff-858b4ba3b588d303d7396271ac3575bbe9585625ee8b278feb1411524d533a97R1-R232): Added new integration tests to verify the private key rotation functionality and ensure the cluster remains accessible after key changes.
* [`tests/unit/test_tls.py`](diffhunk://#diff-a32fba3e75331a440e32977437daaa8d9c93cb87e72dc5f5e3a543794b0df51bR690-R794): Expanded unit tests to cover the new TLS private key configuration and handling.

skourta committed:
fetched lib and added interface in metadata

skourta committed:
added dependencies for the tls lib

skourta committed:
added etcdpaths dataclass

skourta committed:
feat: turn on TLS

skourta committed:
feat: turn off TLS

skourta committed:
add some unit tests and fix failing unit test for the charm

skourta committed:
fix typo

skourta committed:
remove uneeded deps in charmcraft.yaml and update poetry.lock

skourta committed:
add integration tests

skourta committed:
added rolling ops to handle activating and deactivating tls

skourta committed:
enhanced docstrings

skourta committed:
fixed enabling tls on cluster creation

skourta committed:
fixed unit tests and tweaked integration tests

skourta committed:
use 3 units instead of 2 in tls integration tests

skourta committed:
fixed based on feedback

skourta committed:
couple of fixed based on feedback

skourta committed:
added more unit tests

skourta committed:
implementing Mehdi feedback

skourta committed:
fix tests based on feedback

skourta committed:
implement feedback for path handling

skourta committed:
fixed misuse of tenacity

skourta committed:
fix updating certs

skourta committed:
rework to enable single TLS type activation

skourta committed:
update tests

skourta committed:
go from kebab to snake case for relation data

skourta committed:
fix typo

skourta committed:
add logging messagign and some tests

skourta committed:
reformat

skourta committed:
type and bug fix in test

skourta committed:
added unit test for certificate expiration

skourta committed:
Merge branch 'main' into DPE-5962-tls-implement-basic-flow

skourta committed:
reformat files

skourta committed:
Merge main and necessary fixes

skourta committed:
Merge branch 'main' into DPE-5962-tls-implement-basic-flow

skourta committed:
updated tls_certificates library

skourta committed:
remove comment in tests and fix scale up with TLS

skourta committed:
remove comments

skourta committed:
Merge branch 'main' into DPE-5962-tls-implement-basic-flow

skourta committed:
feedback fixes

skourta committed:
feedback fixes

skourta committed:
set private key using juju secrets

carlcsaposs-canonical committed:
Use stage instead of prime in charmcraft files part (#27)

Fixes issue where files in charm directory conflict with files created
by poetry plugin during staging

skourta committed:
swithed to one private key and added integration test

skourta committed:
Merge branch 'main' into DPE-5965-private-keys-juju-secrets

reneradoi committed:
[DPE-6075] Make continuous writes independent of specific units (#19)

The continuous writes function in the integration test suite needs to be
able to run independently. Currently it is executed via the `etcdctl`
command being run in one of the deployed juju units. This is problematic
as we might want to remove this specific unit as part of the scaling/HA
tests.

This PR adjusts integration test setup and installs the deb-package
`etcd-client` on the host machine. The continuous writes function is
then run with this client's `etcdctl` and thus not depending on
execution in specific units anymore.

The function is still kept very simple:
- It needs to be started with `start_continuous_writes(endpoints=...)`
and stopped with `stop_continuous_writes()`.
- It can be started with multiple endpoints though and persists through
changes of Raft or Juju leadership.
- Use the `assert_continuous_writes_increasing` and
`assert_continuous_writes_consistent` helpers to ensure that the writes
continued throughout any cluster/HA related operation.

**Note:**
There is no currently maintained Python client lib for etcd (see
[issue](https://github.com/etcd-io/etcd/issues/13285)), that's why we
keep using `etcdctl` to connect to etcd on each write. Testing with
several of the available, unsupported clients failed because they either
don't support authentication or cannot be installed to due
incompatibilities with protobuf.

skourta committed:
add unit test

skourta committed:
Merge branch 'main' into DPE-5965-private-keys-juju-secrets

skourta committed:
fix integration test after merge

skourta committed:
updated .gitignore

skourta committed:
add marks to the set private key test

skourta committed:
bring num of units back to 3

skourta committed:
[DPE-5966] Implement CA rotation (#26)

This pull request added CA rotation to the charm.

### TLS Certificate Management Enhancements:

* **New TLS CA Rotation States:**
- Added `TLSCARotationState` enum to represent different stages of CA
rotation (`src/literals.py`).

* **CA Rotation and Cleanup Methods:**
- Introduced methods in `TLSManager` for setting CA rotation state,
checking CA status across servers, and cleaning up old CAs
(`src/managers/tls.py`).
[[1]](diffhunk://#diff-a2c036f44ae9bbab595aad84b23bc3106a342c5ab12f7b0cf3d26268b51cdc83R136-R204)
[[2]](diffhunk://#diff-a2c036f44ae9bbab595aad84b23bc3106a342c5ab12f7b0cf3d26268b51cdc83L57-R102)

* **Event Handling Updates:**
- Added `CleanCAEvent` to handle CA cleanup events and integrated it
into the `TLSEvents` class (`src/events/tls.py`).
[[1]](diffhunk://#diff-3dfb4a104b407f7d4b2e07795a63745a09cde228a8dbe86602e7b72ea600fbe4L15-R57)
[[2]](diffhunk://#diff-3dfb4a104b407f7d4b2e07795a63745a09cde228a8dbe86602e7b72ea600fbe4R88-R89)
[[3]](diffhunk://#diff-3dfb4a104b407f7d4b2e07795a63745a09cde228a8dbe86602e7b72ea600fbe4R229-R243)

### Codebase Improvements:

* **Logging Enhancements:**
- Improved logging messages to provide more context during rolling
restarts and CA rotations (`src/charm.py`).
[[1]](diffhunk://#diff-b9ed39bbc9c0387bd3e07da31d13373745534a1cd723d3e292c73496b12e307cL68-R78)
[[2]](diffhunk://#diff-b9ed39bbc9c0387bd3e07da31d13373745534a1cd723d3e292c73496b12e307cR178-R202)

skourta committed:
Merge branch 'main' into DPE-5965-private-keys-juju-secrets

skourta committed:
enhance unit test coverage

skourta committed:
updated snap lib

reneradoi commented on src/events/etcd.py:
```diff
@@ -195,3 +211,35 @@ def update_admin_password(self, admin_secret_id: str) -> None:
+                        self.charm.tls_events.peer_certificate._send_certificate_requests()
+                        self.charm.tls_events.client_certificate._cleanup_certificate_requests()
+                        self.charm.tls_events.client_certificate._send_certificate_requests()
+
```
This whole block of logic should not be our concern. The workflow here should be:
- get secret content
- check for valid format
- call `regenerate_private_key` of the TLS lib

I understand that this is not possible right now as the TLS lib does not support providing an own private key. Unless we wait until [their feature PR](https://github.com/canonical/tls-certificates-interface/pull/300) is available for us, we should at least move this logic out of here and have it in the TLS manager.

Please also add a `todo` comment so that we don't forget to refactor this.

reneradoi commented on tests/integration/tls/test_private_key_rotation.py:
```diff
@@ -0,0 +1,232 @@
+NUM_UNITS = 3
+TEST_KEY = "test_key"
+TEST_VALUE = "42"
+CERTIFICATE_EXPIRY_TIME = 90
```
I believe this can be removed as we don't use it in this test.

reneradoi commented on tests/integration/tls/test_private_key_rotation.py:
```diff
@@ -0,0 +1,232 @@
+@pytest.mark.runner(["self-hosted", "linux", "X64", "jammy", "large"])
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_tls_enabled(ops_test: OpsTest) -> None:
```
This test function seems to a duplicate of the `test_tls_enabled` in `test_tls.py`. Do we need it here for any reason? If not, we can remove it because it does not really matter for testing the rotation of the private key(s).

skourta replied:
My thinking was that we need to validate that TLS is enabled and used before we do anything related to TLS.

reneradoi replied:
ok

reneradoi commented on tests/integration/tls/test_private_key_rotation.py:
```diff
@@ -0,0 +1,232 @@
+    model = ops_test.model_full_name
+    logger.info("Getting current private keys")
+    current_private_keys: list[str] = [
+        (await get_secret_by_label(ops_test, label=f"{LIBID}-private-key-{i}"))["private-key"]  # type: ignore
```
Please add a comment here referencing where the label is generated in the TLS lib's code. If they decide to change that format, our tests might fail and we will go searching.

reneradoi commented on tests/unit/test_charm.py:
```diff
@@ -110,6 +111,27 @@ def test_start():
         state_out = ctx.run(ctx.on.start(), state_in)
         assert state_out.unit_status == ops.BlockedStatus("etcd service not running")
 
+    # non leader waiting promoted
```
Thank you for adding unit test coverage here and for the peer relation events!

reneradoi reviewed (changes requested):
Thank you Smail, I really appreciate the work you did and all the back-and-forth with the TLS team.

Unfortunately the TLS v4 lib has another couple of flaws here. Especially:
- not being able to have separate private keys for multiple use cases
- not being able to provide a user-supplied private key when regenerating it

are major downsides from my point of view. We should only consider the solution implemented here as a workaround, as it introduced potential liabilities to our code when we need to access "protected" methods of the lib to fulfill our use cases.

We as a team should make a decision here if we want to go with the workaround for now or request these features to be delivered by the TLS team and postponing the TLS private key topic until then.

Mehdi-Bendriss commented on src/events/etcd.py:
```diff
@@ -195,3 +211,35 @@ def update_admin_password(self, admin_secret_id: str) -> None:
+        try:
+            if secret_content := get_secret_from_id(self.charm.model, private_key_id).get(
+                "private-key"
+            ):
```
imo, it would be easier to negate on the condition and return to reduce the indentation and nestedness levels here. 
The exceptions are only raised by `get_secret_from_id` i think?

skourta replied:
You are right. I followed the same coding style used for the `update_admin_password`. I will switch it.

Mehdi-Bendriss commented on src/events/etcd.py:
```diff
@@ -195,3 +211,35 @@ def update_admin_password(self, admin_secret_id: str) -> None:
+                    else base64.b64decode(secret_content).decode("utf-8").strip()
+                )
+
+                if not re.match(r"(-+(BEGIN|END) [A-Z ]+-+)", private_key):
```
I believe we should also add this validation on `_config_changed` and put the charm in a blocked state if the validation fails.

skourta replied:
`_on_config_changed` calls this function to update the private key so we get validation in both config changed and secret changed.

Mehdi-Bendriss commented on tests/integration/tls/test_private_key_rotation.py:
```diff
@@ -0,0 +1,232 @@
+            tls_enabled=True,
+        )
+        == TEST_VALUE
+    )
```
nit: missing failure text on assert

Mehdi-Bendriss reviewed (commented):
Thanks Smail! great work! 
My main concern is on the lack of 2 separate private keys, as per the spec.
In addition to this, we should avoid the `# type: ignore` hint, it pollutes the code.

skourta commented on config.yaml:
```diff
@@ -8,4 +8,10 @@ options:
\ No newline at end of file
+      to a secret that contains the following content: `root: <password>`.
+  
+  tls-private-key:
```
Yes, that is how it was initially implemented ([see](https://github.com/canonical/charmed-etcd-operator/pull/29/commits/402f5ae72dc1b341d3526185d28fe0db2367bfae)) then I changed it to one due to the library limits. The TLS team is working on this. 
How about I switch it back to use 2 config options but the underlying secret will be the same until the TLS team PR is merged and the library is updated?

Mehdi-Bendriss replied:
Note: We are going with the fork while the TLS team merges the PR addressing this

skourta commented on src/events/etcd.py:
```diff
@@ -113,11 +122,14 @@ def _on_start(self, event: ops.StartEvent) -> None:
     def _on_config_changed(self, event: ops.ConfigChangedEvent) -> None:
         """Handle config_changed event."""
+        if tls_private_key_id := self.charm.config.get(TLS_PRIVATE_KEY_CONFIG):
+            self.update_private_key(tls_private_key_id)  # type: ignore
```
Because the return type of `get` is `(bool | int | float | str | None)`.

skourta replied:
I removed all of them

Mehdi-Bendriss replied:
In general, if a method returns a value as an indication of a side effect, we should  rely on the return value of the said method

skourta committed:
use 2 config options

skourta committed:
add TLS prefix to status

Mehdi-Bendriss commented on src/events/etcd.py:
```diff
@@ -195,3 +220,60 @@ def update_admin_password(self, admin_secret_id: str) -> None:
+                "private-key"
+            )
+        except (ModelError, SecretNotFoundError) as e:
+            logger.error(e)
+            return
```
This looks risky. This effectively fails silently. I believe we should let the exception propagate and put the charm in a blocked state when this happens -- mainly `on_config_changed`

skourta replied:
@reneradoi this would be relevant for updating the admin password as well.

reneradoi replied:
I agree with Mehdi, in case of failure there needs to be some feedback to the user. This is more difficult for secrets than for an action. Most sane option (without blocking the workload) is to keep everything running and set the unit status to blocked.

For the admin password, I had prepared a PR for this (see [#12](https://github.com/canonical/charmed-etcd-operator/pull/12)). Mykola asked for holding this back until we have the status implementation figured out. His reasoning was, that the blocking status might get overwritten by something else and the user feedback is gone then.

We could keep the logic here as it is for now and update #12 for the private keys, imo.

Mehdi-Bendriss commented on src/events/etcd.py:
```diff
@@ -195,3 +220,60 @@ def update_admin_password(self, admin_secret_id: str) -> None:
+            logger.error(e)
+            return
+
+        if secret_content is None:
```
Under which conditions would `secret_content is None` be true, given we are catching a `SecretNotFoundError` previously?

Mehdi-Bendriss reviewed (commented):
Thanks Smail, I left a couple of comments.

skourta committed:
add status blocked on secret not found and assertions on it in unit tests

skourta committed:
[DPE-5968] integration tests helpers (#31)

This pull request updates multiple test files to replace the
`wait_for_idle` method with the `wait_until` method for better
consistency and reliability in test execution.

Changes in test files:

*
[`tests/integration/ha/test_scaling.py`](diffhunk://#diff-1511afef9daad75af5001b5c55a0fb9dfc6ba38d784615be51dd492e4a330309R18):
Imported `wait_until` from `helpers_deployment` and replaced
`wait_for_idle` with `wait_until` in `test_build_and_deploy` and
`test_scale_up` functions.
[[1]](diffhunk://#diff-1511afef9daad75af5001b5c55a0fb9dfc6ba38d784615be51dd492e4a330309R18)
[[2]](diffhunk://#diff-1511afef9daad75af5001b5c55a0fb9dfc6ba38d784615be51dd492e4a330309L42-R43)
[[3]](diffhunk://#diff-1511afef9daad75af5001b5c55a0fb9dfc6ba38d784615be51dd492e4a330309L63-R65)

*
[`tests/integration/test_charm.py`](diffhunk://#diff-08291d7f4ad7b2caecacb6ae26af540ebbaede9fe42d5e8e1b39d2bdefe4c2d4R20):
Imported `wait_until` from `helpers_deployment` and replaced
`wait_for_idle` with `wait_until` in `test_build_and_deploy` and
`test_update_admin_password` functions.
[[1]](diffhunk://#diff-08291d7f4ad7b2caecacb6ae26af540ebbaede9fe42d5e8e1b39d2bdefe4c2d4R20)
[[2]](diffhunk://#diff-08291d7f4ad7b2caecacb6ae26af540ebbaede9fe42d5e8e1b39d2bdefe4c2d4L41-R42)
[[3]](diffhunk://#diff-08291d7f4ad7b2caecacb6ae26af540ebbaede9fe42d5e8e1b39d2bdefe4c2d4L98-R99)
[[4]](diffhunk://#diff-08291d7f4ad7b2caecacb6ae26af540ebbaede9fe42d5e8e1b39d2bdefe4c2d4L107-R108)

*
[`tests/integration/tls/test_ca_rotation.py`](diffhunk://#diff-a8362a8580830e574ae0f1e2fbd35adbf5ea51ed2ae3a85e78c5a2eebaa4d0c7R25):
Imported `wait_until` from `helpers_deployment` and replaced
`wait_for_idle` with `wait_until` in `test_build_and_deploy_with_tls`
function.
[[1]](diffhunk://#diff-a8362a8580830e574ae0f1e2fbd35adbf5ea51ed2ae3a85e78c5a2eebaa4d0c7R25)
[[2]](diffhunk://#diff-a8362a8580830e574ae0f1e2fbd35adbf5ea51ed2ae3a85e78c5a2eebaa4d0c7L56-R57)
[[3]](diffhunk://#diff-a8362a8580830e574ae0f1e2fbd35adbf5ea51ed2ae3a85e78c5a2eebaa4d0c7L131-R133)

*
[`tests/integration/tls/test_private_key_rotation.py`](diffhunk://#diff-858b4ba3b588d303d7396271ac3575bbe9585625ee8b278feb1411524d533a97R32):
Imported `wait_until` from `helpers_deployment` and replaced
`wait_for_idle` with `wait_until` in `test_build_and_deploy_with_tls`
and `test_set_private_key` functions.
[[1]](diffhunk://#diff-858b4ba3b588d303d7396271ac3575bbe9585625ee8b278feb1411524d533a97R32)
[[2]](diffhunk://#diff-858b4ba3b588d303d7396271ac3575bbe9585625ee8b278feb1411524d533a97L63-R64)
[[3]](diffhunk://#diff-858b4ba3b588d303d7396271ac3575bbe9585625ee8b278feb1411524d533a97L170-R171)
[[4]](diffhunk://#diff-858b4ba3b588d303d7396271ac3575bbe9585625ee8b278feb1411524d533a97L235-R236)

*
[`tests/integration/tls/test_tls.py`](diffhunk://#diff-5877b0903d94204b2a76f3b252020e23fc436bd77688fe77e00a66b864ac9fe0R25):
Imported `wait_until` from `helpers_deployment` and replaced
`wait_for_idle` with `wait_until` in multiple functions including
`test_build_and_deploy_with_tls`, `test_disable_tls`, `test_enable_tls`,
`test_disable_and_enable_peer_tls`,
`test_disable_and_enable_client_tls`, and `test_certificate_expiration`.
[[1]](diffhunk://#diff-5877b0903d94204b2a76f3b252020e23fc436bd77688fe77e00a66b864ac9fe0R25)
[[2]](diffhunk://#diff-5877b0903d94204b2a76f3b252020e23fc436bd77688fe77e00a66b864ac9fe0L57-R58)
[[3]](diffhunk://#diff-5877b0903d94204b2a76f3b252020e23fc436bd77688fe77e00a66b864ac9fe0L119-R120)
[[4]](diffhunk://#diff-5877b0903d94204b2a76f3b252020e23fc436bd77688fe77e00a66b864ac9fe0L175-R176)
[[5]](diffhunk://#diff-5877b0903d94204b2a76f3b252020e23fc436bd77688fe77e00a66b864ac9fe0L246-R247)
[[6]](diffhunk://#diff-5877b0903d94204b2a76f3b252020e23fc436bd77688fe77e00a66b864ac9fe0L302-R303)
[[7]](diffhunk://#diff-5877b0903d94204b2a76f3b252020e23fc436bd77688fe77e00a66b864ac9fe0L371-R372)
[[8]](diffhunk://#diff-5877b0903d94204b2a76f3b252020e23fc436bd77688fe77e00a66b864ac9fe0L423-R424)
[[9]](diffhunk://#diff-5877b0903d94204b2a76f3b252020e23fc436bd77688fe77e00a66b864ac9fe0L490-R491)
[[10]](diffhunk://#diff-5877b0903d94204b2a76f3b252020e23fc436bd77688fe77e00a66b864ac9fe0L549-R550)

skourta committed:
upgrade lib from fork

skourta committed:
add checks on client/peer secret nto changing when the other changes

skourta committed:
Merge branch '3.5/edge' into DPE-5965-private-keys-juju-secrets

skourta committed:
added TODO comment

skourta committed:
update dp interfaces

skourta committed:
fix flakey unit test

skourta committed:
remove build charm

Mehdi-Bendriss commented on config.yaml:
```diff
@@ -8,4 +8,14 @@ options:
+  tls-client-private-key:
+    type: secret
+    description: |
+      A Juju secret URI of a secret containing the private key for server-to-client TLS certificates.
```
```suggestion
      A Juju secret URI of a secret containing the private key for client-to-server TLS certificates.
```

Mehdi-Bendriss reviewed (dismissed):
Thanks Smail! Great work! (just a small typo to fix)

skourta committed:
fix typo

Co-authored-by: Mehdi Bendriss <bendrissmehdi@gmail.com>

skourta committed:
add empty line in config.yaml

reneradoi reviewed (approved):
Thank you Smail!
